features/237-cms-validation/api.md

```diff
@@ -0,0 +1,146 @@
+# CMS Validation
+
+## Requirements
+
+- Prevent cms user from entering invalid data.
+- Allow developer to use common and customisable validation in cms config on field level and form level.
+- Display error message, which can be defined in config, for invalid data.
+
+## Common Use Cases
+
+- Non empty text input.
+- Text input with max length.
+- Text input with certain pattern, e.g. credit card number.
+- Number or datetime input value within range.
+- One field cannot be empty if another field is not empty.
+
+## Overview
+
+- Validation is a process that evaluate data with an expression, the result is either success or failure.
+- Validation can be defined in record edit page and new page in cms config.
+- Validation happens after use click save button and before all server api calls are fired.
+- If any validation result is failure, error message should be displayed and no server api calls should be fired.
+- Error message of field validation appears next / below the field, while that of form validation appears below the form.
+- Field validation is executed before form validation.
+
+## Configuration
+
+- One validation item is a dictionary which contains the validation configuration (depends on where it is placed).
+  - Each validation optionally contains a `message` field, which would be the error message if that validation item cannot pass.
+  - Otherwise, a default error message would be shown.
+- `validation` may be a validation item or contains a list of validation items.
```

Comment:
does it meant, to configure multiple validation for the same field, it should be like...

```
new:
  fields:
    - name: xxx
       type: xxx
       validation:
         validation:
           pattern: /^xxx$/
         validation:
           required: true
```

If that's the case, would it be better call it `validations`?

Reply:
Configuration for more than one validation.

```
fields:
- name: xxx
   type: xxx
   validation:
     - expression: xxx
     - expression: xxx
```

Reply:
@Steven-Chan so that will also be:

```
validation:
  - pattern: /^xxx$/
  - required: true
  - expression: xxx
```
?

BTW that should be `validations`?

Reply:
You're right, like other config keys `fields`, `actions`, they are plural and only accept array. For consistency, I will update the configuration here to follow other config.